torch/_functorch/aot_autograd.py

```diff
@@ -138,86 +138,365 @@ def posthook(grad_input, grad_output):
-    tensor_meta: Optional[Tensor]
-
-# This class tells us about how to perform a metadata mutation on forward inputs.
-# it only applies to forward inputs that experience metadata-only mutations
+    base_idx: Optional[int]
+    # Only important when output_type == OutputType.alias_of_input
+    # Tells the runtime code to return args[base_idx] *exactly*, and not generate a view off of it.
+    is_input_tensor: bool
+    # Only important when output_type == OutputType.alias_of_intermediate
+    # Tells the code that traces the joint that we should append output._base as an output
+    # in the forward graph.
+    add_base_to_outputs: bool
+
+# This class tells us info about user inputs.
 @dataclass(frozen=True)
 class InputAliasInfo:
-    # This object gives us information about how to perform a metadata-mutation
-    # on original_fwd_inputs[base_idx]
-    #   (This is an index into the inputs *before* we make synthetic bases)
-    base_idx: int
-    # sizes, strides and storage offset of the aliased output are all returned as actual (sym)ints
-    # in the compiled forward. These indices tell us where in the forward outputs to grab them.
-    sizes_idx: int
-    strides_idx: int
-    storage_offset_idx: int
-    # We store the actual output alias that we traced in the forward (should be a fake tensor)
-    # to grab any other non-symbolic properties on the output alias, like requires_grad.
-    tensor_meta: Tensor
+    mutation_type: MutationType
 
 # This class encapsulates all aliasing + mutation info we need about the forward graph
 # See a more detailed overview of the edge case handling at
 # https://docs.google.com/document/d/19UoIh_SVrMy_b2Sx5ZaeOJttm6P0Qmyss2rdBuyfoic/edit
 @dataclass(frozen=True)
 class ViewAndMutationMeta:
-    # length: # user forward inputs
-    # For every input, tells us whether the input:
-    # (a) is not mutated
-    # (b) only metadata is mutated
-    # (c) data (and maybe metadta) is mutated
-    mutated_input_info: List[MutationType]
-    # length: (# inputs of the user forward)
-    # metadata_mutation_input_info[i] is not None <====> mutated_input_info[i] == MutationType.metadata_only
-    # We stash the updated FakeTensor that we traced with in the forward in here,
-    # that way we can use it to replay the metadata mutation
-    metadata_mutation_input_info: List[Optional[InputAliasInfo]]
-    # length: # outputs in the compiled forward (not including output alias symints). Equal to:
-    # length: (# inputs w data mutations) + (# outputs that don't alias inputs)
+    # length = # user inputs
+    # This gives us info about every input, and what sort of mutation happened to it (if any)
+    input_info: List[InputAliasInfo]
+    # length = # user outputs + # intermediates that are returned as graph outputs
+    # This gives us info about every output (mostly around whether it aliases other tensors)
+    output_info: List[OutputAliasInfo]
+    # length = # outputs in the compiled forward. Equal to:
```

Comment:
The formatting here is confusing because you might accidentally think that the length applies to the list above, but it applies below. Add some newlines separating the comment+field will help